### DISCUSSION TITLE
Inter-process communication

### DISCUSSION THREAD

**Discussion by andreacavagna01** [Noovolari/leapp#53] [Q&A] [answered]
What is the best way to implement Inter-process communication through a backend and a GUI for a Desktop application in Electron?

Any thoughts @urz9999?

**Comment by giovannibagolin** [comment]
Based on my knowledge I would say the best way is to implement a messages queue. An example can be found here: https://www.geeksforgeeks.org/ipc-using-message-queues/

**Reply by urz9999** [reply]
yeah, to keep things as standardised as possibile and also as simple as possibile we are thinking about gRPC via libraries for specific languages, what do you think?

**Reply by ericvilla** [reply]
Hi @gbagolin! Thanks for the reply and for sharing the link. What about IPC in terms of cross-platform support?

**Reply by giovannibagolin** [reply]
Hi @ericvilla, @urz9999, thanks for your review. 
Given the requirement of a cross-platform application, and after having read quickly what the "g" in gRPC standed for, I guess this fits well as solution to your problem.

**Comment by ericvilla** [comment, marked as answer]
That's an important point and I would add detail to the question: inter-process communication should work on Mac OS, Windows, and Linux. 

As far as now, our idea is to opt for a technology that can be re-used on the listed platforms. For what concerns Unix environments, Unix Sockets are well suited for inter-process communication, but they do not cover Windows; the same happens for Windows' Named Pipes, that do not cover Unix systems. 

We started exploring the usage of an HTTP server as "back-end" of the application or a gRPC server. So, we can definitively implement this type of software architecture; we just need to choose the right technology to do that. We have to consider the language; to choose the right one we have to investigate if there are robust libraries support for the interprocess protocol we are intended to adopt. 

When it comes to choose the right language, we might have to take into account the pros and cons of an Interpreted or Compiled language. I saw many plugin-based solutions that are based on a Compiled language.

I hope it makes sense to you, and that we can find the right path to implement!
